File: packages/web/src/algorithms/getNucleotideComposition.ts
export function getNucleotideComposition(alignedQuery: string): Record<string, number> {
  const result: Record<string, number> = {}
  for (let i = 0; i < alignedQuery.length; i++) {
    const char = alignedQuery[i]
    if (result[char] === undefined) {
      result[char] = 1
    } else {
      result[char]++
    }
  }
  return result
}

File: packages/nextclade/src/analyze/getNucleotideComposition.cpp
#include "getNucleotideComposition.h"

#include <nextclade/nextclade.h>

#include <map>

#include "../utils/mapFind.h"
#include "../utils/safe_cast.h"

namespace Nextclade {
  std::map<Nucleotide, int> getNucleotideComposition(const NucleotideSequence& alignedQuery) {
    std::map<Nucleotide, int> result;
    const auto& length = safe_cast<int>(alignedQuery.size());
    for (int i = 0; i < length; ++i) {
      const auto& nuc = alignedQuery[i];
      result[nuc] = ++(result[nuc]);
    }
    return result;
  }
}// namespace Nextclade
